### PR TITLE
[search] Implemented gradient descent for a linear model estimation.

### PR DESCRIPTION
**Pull request by ygorshenin**
@mpimenov PTAL

### CODE REVIEW

File: search/search_query.cpp [hunk 1]
```diff
@@ -661,9 +661,22 @@ class PreResult2Maker
...
       if (!ft.GetName(lang, name))
         continue;
-      auto score = GetNameScore(name, m_params, preInfo.m_startToken, preInfo.m_endToken);
+      vector<strings::UniString> tokens;
+      SplitUniString(NormalizeAndSimplifyString(name), MakeBackInsertFunctor(tokens), Delimiters());
+
+      auto score = GetNameScore(tokens, m_params, preInfo.m_startToken, preInfo.m_endToken);
+      auto coverage =
+          tokens.empty() ? 0 : static_cast<double>(preInfo.m_endToken - preInfo.m_startToken) /
+                                   static_cast<double>(tokens.size());
```
Comment (mpimenov) [inline search/search_query.cpp]: clang-format?

File: search/v2/ranking_info.hpp [hunk 2]
```diff
@@ -25,6 +25,9 @@ struct RankingInfo
   // Score for the feature's name.
   NameScore m_nameScore = NAME_SCORE_ZERO;
 
+  // Number of tokens from the query matched to a feature name.
```
Comment (mpimenov) [inline search/v2/ranking_info.hpp]: s/Number/Fraction/

File: search/search_query.cpp (hunk 1 above)
Comment (ygorshenin) [inline search/search_query.cpp]: This is clang-format

File: search/search_quality/scoring_model.py [hunk 3]
```diff
@@ -66,20 +45,139 @@ def compute_ndcg(scores):
...
+        relevances = relevances[scores.argsort()[::-1]]
+        ndcgs.append(compute_ndcg(relevances))
+
+    return np.array(ndcgs)
+
+
+def gradient_descent(w_init, grad, eps=1e-6, lam=1e-3, num_steps=1000):
+    n = len(w_init)
+    w, dw = np.copy(w_init), np.zeros(n)
+    for step in range(1, num_steps):
```
Comment (mpimenov) [inline search/search_quality/scoring_model.py]: ```for step in range(num_steps):```

File: search/v2/ranking_info.hpp (hunk 2 above)
Comment (ygorshenin) [inline search/v2/ranking_info.hpp]: Done

File: search/search_quality/scoring_model.py [hunk 4]
```diff
@@ -66,20 +45,139 @@ def compute_ndcg(scores):
...
+
+class RankingSVMLoss:
+    """
+    Represents a loss function with a gradient for a RankingSVM model.
+    Simple version of a loss function for a ranked list of features
+    has following form:
+
+    loss(w) = sum{i, j: max(0, 1 - sign(y[j] - y[i]) * dot(w, x[j] - x[i]))} + lam * dot(w, w)
+
+    This version is slightly modified, as we dealing with a group of
```
Comment (mpimenov) [inline search/search_quality/scoring_model.py]: s/we/we are/

File: search/search_quality/scoring_model.py (hunk 3 above)
Comment (ygorshenin) [inline search/search_quality/scoring_model.py]: Done

File: search/search_quality/scoring_model.py (hunk 4 above)
Comment (ygorshenin) [inline search/search_quality/scoring_model.py]: Done

File: search/search_quality/scoring_model.py [hunk 5]
```diff
@@ -66,20 +45,139 @@ def compute_ndcg(scores):
...
+        return g
+
+
+class RankingSVMLoss:
+    """
+    Represents a loss function with a gradient for a RankingSVM model.
+    Simple version of a loss function for a ranked list of features
+    has following form:
+
+    loss(w) = sum{i, j: max(0, 1 - sign(y[j] - y[i]) * dot(w, x[j] - x[i]))} + lam * dot(w, w)
```
Comment (mpimenov) [inline search/search_quality/scoring_model.py]: ```sum{i < j}```?

File: search/search_quality/scoring_model.py [hunk 6]
```diff
@@ -66,20 +45,139 @@ def compute_ndcg(scores):
...
+    normalize_data(data)
 
+    best_w = np.ones(len(FEATURES))
+    best_mean = np.mean(compute_ndcg_for_w(data, best_w))
+
+    loss = RankingSVMLoss(data, lam=1e-3)
+    grad = lambda w: loss.gradient(w)
+
+    num_steps = 1000
+    for i in range(1, num_steps + 1):
```
Comment (mpimenov) [inline search/search_quality/scoring_model.py]: s/1, //

Comment (ygorshenin) [inline search/search_quality/scoring_model.py]: Не, проще просто range(num_steps)

**Comment by mpimenov** [comment]
LGTM after fixes
